File: realm/src/main/java/xyz/truenight/databinding/realm/RealmObservableList.java
package xyz.truenight.databinding.realm;

import android.databinding.ListChangeRegistry;
import android.databinding.ObservableList;

import java.util.AbstractList;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

import io.realm.OrderedCollectionChangeSet;
import io.realm.OrderedRealmCollection;
import io.realm.OrderedRealmCollectionChangeListener;
import io.realm.Realm;
import io.realm.RealmConfiguration;
import io.realm.RealmList;
import io.realm.RealmModel;
import io.realm.RealmQuery;
import io.realm.RealmResults;

/**
 * An {@link ObservableList} that uses {@link OrderedRealmCollectionChangeListener} to calculate and dispatch it's change
 * updates.
 */
public class RealmObservableList<T extends RealmModel> extends AbstractList<T> implements ObservableList<T> {

    public static <E extends RealmModel> OnApplyQuery<E> all() {
        //noinspection unchecked
        return RealmObservableField.all();
    }

    private final Class<T> cls;
    private final OnApplyQuery<T> applyQuery;
    private final RealmConfiguration realmConfig;
    private Realm realm;
    private OrderedRealmCollection<T> items;
    private final OrderedRealmCollectionChangeListener listener;
    private final ListChangeRegistry listeners = new ListChangeRegistry();
    private List<T> mirror;
    private boolean updateOnModification;
    private FindFunction<T> findFunction;

    /**
     * Creates a new AutoQueryRealmObservableList of type T.
     *
     * @param apply Realm query.
     */
    public RealmObservableList(Class<T> cls, OnApplyQuery<T> apply) {
        this(cls, apply, null);
    }

    /**
     * Creates a new AutoQueryRealmObservableList of type T.
     *
     * @param apply Realm query.
     */
    public RealmObservableList(RealmConfiguration configuration, Class<T> cls, OnApplyQuery<T> apply) {
        this(configuration, cls, apply, null);
    }

    /**
     * Creates a new AutoQueryRealmObservableList of type T.
     *
     * @param apply Realm query.
     */
    public RealmObservableList(Class<T> cls, OnApplyQuery<T> apply, FindFunction<T> function) {
        this(cls, apply, function, true);
    }

    /**
     * Creates a new AutoQueryRealmObservableList of type T.
     *
     * @param apply Realm query.
     */
    public RealmObservableList(RealmConfiguration configuration, Class<T> cls, OnApplyQuery<T> apply, FindFunction<T> function) {
        this(configuration, cls, apply, function, true);
    }

    public RealmObservableList(Class<T> cls, OnApplyQuery<T> apply, FindFunction<T> function, boolean updateOnModification) {
        this(Realm.getDefaultConfiguration(), cls, apply, function, updateOnModification);
    }

    /**
     * Creates a new AutoQueryRealmObservableList of type T.
     *
     * @param apply Realm query.
     */
    public RealmObservableList(RealmConfiguration configuration, Class<T> cls, OnApplyQuery<T> apply, FindFunction<T> function, boolean updateOnModification) {
        this.realmConfig = configuration;
        this.cls = cls;
        this.applyQuery = apply;
        this.findFunction = function;
        this.mirror = Collections.emptyList();
        this.listener = createListener();
        this.updateOnModification = updateOnModification;
    }

    @Deprecated
    @Override
    public T set(int index, T element) {
        return super.set(index, element);
    }

    @Deprecated
    @Override
    public void clear() {
        super.clear();
    }

    @Deprecated
    @Override
    public boolean add(T t) {
        return super.add(t);
    }

    @Deprecated
    @Override
    public void add(int index, T element) {
        super.add(index, element);
    }

    @Deprecated
    @Override
    public boolean addAll(Collection<? extends T> c) {
        return super.addAll(c);
    }

    @Deprecated
    @Override
    public boolean addAll(int index, Collection<? extends T> c) {
        return super.addAll(index, c);
    }

    @Deprecated
    @Override
    protected void removeRange(int fromIndex, int toIndex) {
        super.removeRange(fromIndex, toIndex);
    }

    @Deprecated
    @Override
    public boolean remove(Object o) {
        return super.remove(o);
    }

    @Deprecated
    @Override
    public boolean removeAll(Collection<?> c) {
        return super.removeAll(c);
    }

    @Deprecated
    @Override
    public T remove(int index) {
        return super.remove(index);
    }

    @Deprecated
    @Override
    public boolean retainAll(Collection<?> c) {
        return super.retainAll(c);
    }

    private OrderedRealmCollectionChangeListener createListener() {
        return new OrderedRealmCollectionChangeListener() {

            // mirror list state and apply new only after onChange called
            @Override
            public void onChange(Object collection, OrderedCollectionChangeSet changeSet) {
                mirror = new ArrayList<>(items);
                // null Changes means the async query returns the first time.
                if (changeSet == null) {
                    listeners.notifyChanged(RealmObservableList.this);
                    return;
                }
                // For deletions, the adapter has to be notified in reverse order.
                OrderedCollectionChangeSet.Range[] deletions = changeSet.getDeletionRanges();
                for (int i = deletions.length - 1; i >= 0; i--) {

                    OrderedCollectionChangeSet.Range range = deletions[i];
                    modCount += 1;
                    listeners.notifyRemoved(RealmObservableList.this, range.startIndex, range.length);
                }

                OrderedCollectionChangeSet.Range[] insertions = changeSet.getInsertionRanges();
                for (OrderedCollectionChangeSet.Range range : insertions) {
                    modCount += 1;
                    listeners.notifyInserted(RealmObservableList.this, range.startIndex, range.length);
                }

                if (!updateOnModification) {
                    return;
                }

                OrderedCollectionChangeSet.Range[] modifications = changeSet.getChangeRanges();
                for (OrderedCollectionChangeSet.Range range : modifications) {
                    listeners.notifyChanged(RealmObservableList.this, range.startIndex, range.length);
                }
            }
        };
    }

    @Override
    public void addOnListChangedCallback(OnListChangedCallback<? extends ObservableList<T>> listener) {
        boolean first = listeners.isEmpty();
        listeners.add(listener);
        if (first) {
            realm = Realm.getInstance(realmConfig);
            RealmQuery<T> realmQuery = applyQuery.onApply(realm.where(cls));
            if (findFunction == null) {
                items = realmQuery.findAllAsync();
            } else {
                items = findFunction.find(realmQuery);
            }
            mirror = new ArrayList<>(items);
            addRealmChangeListener();
            listeners.notifyChanged(RealmObservableList.this);
        }
    }

    @Override
    public void removeOnListChangedCallback(OnListChangedCallback<? extends ObservableList<T>> listener) {
        listeners.remove(listener);
        if (listeners.isEmpty()) {
            removeRealmChangeListener();
            items = null;
            mirror = Collections.emptyList();
            realm.close();
        }
    }

    private void addRealmChangeListener() {
        if (items instanceof RealmResults) {
            RealmResults<T> results = (RealmResults<T>) items;
            //noinspection unchecked
            results.addChangeListener(listener);
        } else if (items instanceof RealmList) {
            RealmList<T> list = (RealmList<T>) items;
            //noinspection unchecked
            list.addChangeListener(listener);
        } else {
            throw new IllegalArgumentException("RealmCollection not supported: " + items.getClass());
        }
    }

    private void removeRealmChangeListener() {
        if (items instanceof RealmResults) {
            RealmResults<T> results = (RealmResults<T>) items;
            //noinspection unchecked
            results.removeChangeListener(listener);
        } else if (items instanceof RealmList) {
            RealmList<T> list = (RealmList<T>) items;
            //noinspection unchecked
            list.removeChangeListener(listener);
        } else {
            throw new IllegalArgumentException("RealmCollection not supported: " + items.getClass());
        }
    }

    @Override
    public T get(int i) {
        return mirror.get(i);
    }

    @Override
    public int size() {
        return mirror.size();
    }

}

File: realm/src/main/java/xyz/truenight/databinding/realm/RealmObservableList.kt
package xyz.truenight.databinding.realm

import androidx.databinding.ListChangeRegistry
import androidx.databinding.ObservableList
import io.realm.*
import java.util.*
import java.util.concurrent.CopyOnWriteArraySet

/**
 * An [ObservableList] that uses [OrderedRealmCollectionChangeListener] to calculate and dispatch it's change
 * updates.
 */
class RealmObservableList<T : RealmModel> private constructor(private val realmConfig: RealmConfiguration, private val cls: Class<T>, private val query: (RealmQuery<T>) -> RealmResults<T>, private val updateOnModification: Boolean) : AbstractList<T>(), ObservableList<T> {

    private val count = CopyOnWriteArraySet<ObservableList.OnListChangedCallback<*>>()
    private var realm: Realm? = null
    private var items: RealmResults<T>? = null
    private val listener: OrderedRealmCollectionChangeListener<RealmResults<T>> = createListener()
    private val listeners = ListChangeRegistry()
    private var mirror: List<T> = emptyList()

    @Deprecated("Method not supported")
    override fun set(index: Int, element: T): T {
        return super.set(index, element)
    }

    @Deprecated("Method not supported")
    override fun clear() {
        super.clear()
    }

    @Deprecated("Method not supported")
    override fun add(t: T): Boolean {
        return super.add(t)
    }

    @Deprecated("Method not supported")
    override fun add(index: Int, element: T) {
        super.add(index, element)
    }

    @Deprecated("Method not supported")
    override fun addAll(c: Collection<T>): Boolean {
        return super.addAll(c)
    }

    @Deprecated("Method not supported")
    override fun addAll(index: Int, c: Collection<T>): Boolean {
        return super.addAll(index, c)
    }

    @Deprecated("Method not supported")
    override fun removeRange(fromIndex: Int, toIndex: Int) {
        super.removeRange(fromIndex, toIndex)
    }

    @Deprecated("Method not supported")
    override fun remove(element: T?): Boolean {
        return super.remove(element)
    }

    @Deprecated("Method not supported")
    override fun removeAll(c: Collection<T>): Boolean {
        return super.removeAll(c)
    }

    @Deprecated("Method not supported")
    override fun removeAt(index: Int): T {
        return super.removeAt(index)
    }

    @Deprecated("Method not supported")
    override fun retainAll(elements: Collection<T>): Boolean {
        return super.retainAll(elements)
    }

    private fun createListener(): OrderedRealmCollectionChangeListener<RealmResults<T>> {
        return object : OrderedRealmCollectionChangeListener<RealmResults<T>> {

            // mirror list state and apply new only after onChange called
            override fun onChange(collection: RealmResults<T>, changeSet: OrderedCollectionChangeSet) {
                mirror = ArrayList(items!!)
                // null Changes means the async query returns the first time.
                if (changeSet == null) {
                    listeners.notifyChanged(this@RealmObservableList)
                    return
                }
                // For deletions, the adapter has to be notified in reverse order.
                val deletions = changeSet.deletionRanges
                for (i in deletions.indices.reversed()) {

                    val range = deletions[i]
                    modCount += 1
                    listeners.notifyRemoved(this@RealmObservableList, range.startIndex, range.length)
                }

                val insertions = changeSet.insertionRanges
                for (range in insertions) {
                    modCount += 1
                    listeners.notifyInserted(this@RealmObservableList, range.startIndex, range.length)
                }

                if (!updateOnModification) {
                    return
                }

                val modifications = changeSet.changeRanges
                for (range in modifications) {
                    listeners.notifyChanged(this@RealmObservableList, range.startIndex, range.length)
                }
            }
        }
    }

    override fun addOnListChangedCallback(listener: ObservableList.OnListChangedCallback<out ObservableList<T>>) {
        listeners.add(listener)
        if (!EXCLUDE.contains(listener.javaClass.name)) {
            if (count.isEmpty()) {
                val realm = Realm.getInstance(realmConfig)
                this.realm = realm
                val results = query(realm.where(cls))

                mirror = ArrayList(items)
                results.addChangeListener(this.listener)
                items = results
                listeners.notifyChanged(this@RealmObservableList)
            }
            count.add(listener)
        }
    }

    override fun removeOnListChangedCallback(listener: ObservableList.OnListChangedCallback<out ObservableList<T>>) {
        listeners.remove(listener)
        count.remove(listener)
        if (count.isEmpty()) {
            items?.removeChangeListener(this.listener)
            items = null
            mirror = emptyList()
            realm?.close()
        }
    }

    override fun get(index: Int): T {
        return mirror[index]
    }

    override val size: Int
        get() = mirror.size

    companion object {

        private val EXCLUDE = HashSet<String>()

        fun excludeCallback(cls: Class<out ObservableList.OnListChangedCallback<*>>) {
            EXCLUDE.add(cls.name)
        }

        fun removeExclusion(cls: Class<out ObservableList.OnListChangedCallback<*>>) {
            EXCLUDE.remove(cls.name)
        }

        fun excludeCallback(className: String) {
            EXCLUDE.add(className)
        }

        fun removeExclusion(className: String) {
            EXCLUDE.remove(className)
        }

        inline fun <reified T : RealmModel> create(
                realmConfig: RealmConfiguration = Realm.getDefaultConfiguration()!!,
                noinline query: (RealmQuery<T>) -> RealmResults<T>,
                updateOnModification: Boolean = true
        ) = create(realmConfig, T::class.java, query, updateOnModification)

        @JvmStatic
        @JvmOverloads
        fun <T : RealmModel> create(
                realmConfig: RealmConfiguration = Realm.getDefaultConfiguration()!!,
                cls: Class<T>, query: (RealmQuery<T>) -> RealmResults<T>,
                updateOnModification: Boolean = true
        ) = RealmObservableList(realmConfig, cls, query, updateOnModification)
    }

}